src/aof.c

```diff
@@ -313,8 +710,8 @@ int startAppendOnly(void) {
             serverLog(LL_WARNING,"AOF was enabled but there is already an AOF rewriting in background. Stopping background AOF and starting a rewrite now.");
             killAppendOnlyChild();
         }
-        if (rewriteAppendOnlyFileBackground() == C_ERR) {
-            close(newfd);
+
+        if (rewriteAppendOnlyFileBackground(1) == C_ERR) {
```

Comment:
Okay, I have made aof open before redis-fork so that we can find disk errors in advance.

Reply:
i see that now `startAppendonly` creates a new file, but it doesn't write to `server.aof_fd` until after `rewriteAppendOnlyFileBackground` (same as it is in unstable), however, rewriteAppendOnlyFileBackground does this:
```c
if (server.aof_fd != -1) openNewIncrAofForAppend();
```
so i think openNewIncrAofForAppend still has a chance to `exit`.
and in any case, it'll open the file again, writing to `server.aof_fd`, and we probably have an FD leak.

Reply:
Sorry I didn't understand what you mean. When we  calls `rewriteAppendOnlyFileBackground` in `startAppendonly` function, `server.aof_fd` must be -1 now, so 
 `rewriteAppendOnlyFileBackground` will not be opened new aof file. I don’t understand under what circumstances the FD leak you mentioned happened。

Reply:
ohh, i now see i got the condition in rewriteAppendOnlyFileBackground wrong (the one in step 4 below)

1. startAppendOnly does `newfd = open()`
2. startAppendOnly calls rewriteAppendOnlyFileBackground
3. rewriteAppendOnlyFileBackground does `if (server.aof_fd != -1) openNewIncrAofForAppend();`
4. openNewIncrAofForAppend does `server.aof_fd = open()`
5. startAppendOnly does `server.aof_fd = newfd;` (after rewriteAppendOnlyFileBackground returns with success)

i'd still argue that this is a bit confusing, and that maybe it would have been better to open the new file just in one place.
i understand that the `open()` in `startAppendOnly` is the one i asked for (for graceful failure). and that we'll need one for reoccurring rewrite (which can be in rewriteAppendOnlyFileBackground).

maybe we can solve it in a cleaner way if we move the opening of the new file in rewriteAppendOnlyFileBackground to before `fork()` is called (it doesn't really matters if it is done in the parent before or after fork), and then maybe it's nicer to have just one call to `open`, or even two in some if-else chain, and it'll be less confusing as to which function is responsible of opening the new file and which one is responsible of setting `server.aof_fd` (they'll all be in just one place).

Reply:
I think this has nothing to do with whether move the `openNewIncrAofForAppend` before fork, because the `rewriteAppendOnlyFileBackground` function is not only called in `startAppendOnly`, but also in `bgrewriteaofCommand` and `aof_rewrite_scheduled` and automatic rewriting in server cron.

So `openNewIncrAofForAppend` handles common things at runtime: once AOFRW, it opens a new file, and if it fails, it exits the process. Even if we move it to the front of the fork, we still have to do this (mean exit).

The reason why `startAppendOnly` is so special is that we want to give users a chance to know the disk problem in advance, instead of exiting directly. But I think this should not affect the internal implementation of `rewriteAppendOnlyFileBackground`.

Currently, the implementation of rewriteAppendOnlyFileBackground is like this(deleted the irrelevant code):
```
    if ((childpid = redisFork(CHILD_TYPE_AOF)) == 0) {
        
    } else {
      
        serverLog(LL_NOTICE,
            "Background append only file rewriting started by pid %ld",(long) childpid);
        server.aof_rewrite_scheduled = 0;
        server.aof_rewrite_time_start = time(NULL);

        /* We set aof_selected_db to -1 in order to force the next call to the
         * feedAppendOnlyFile() to issue a SELECT command, so the differences
         * accumulated by the parent into server.aof_rewrite_buf will start
         * with a SELECT statement and it will be safe to merge. */
        server.aof_selected_db = -1;
        replicationScriptCacheFlush();
        flushAppendOnlyFile(1);
        /* Only open new INCR AOF when AOF enabled. */
        if (server.aof_fd != -1) openNewIncrAofForAppend();
        return C_OK;
    }
```